File: solutions/typescript/allergies/1/allergies.ts
const ALLERGENS = [
  'eggs',
  'peanuts',
  'shellfish',
  'strawberries',
  'tomatoes',
  'chocolate',
  'pollen',
  'cats',
] as const;

type Allergen = (typeof ALLERGENS)[number];

export class Allergies {
  allergies: Allergen[];

  constructor(allergenIndex: number) {
    this.allergies = ALLERGENS.filter(
      (_, index) => allergenIndex & (1 << index)
    );
  }

  public list(): Allergen[] {
    return this.allergies;
  }

  public allergicTo(allergen: Allergen): boolean {
    return this.allergies.includes(allergen);
  }
}

File: solutions/r/allergies/1/allergies.R
allergy <- function(num) {
  allergies <- bitwAnd(num, 2^(0:7))
  names(allergies) <- c("eggs", "peanuts", "shellfish", "strawberries", 
                        "tomatoes", "chocolate", "pollen", "cats")
  allergies
}

allergic_to <- function(allergy_object, allergy) {
  allergy_object[allergy] > 0
}

list_allergies <- function(allergy_object) {
  names(which(allergy_object > 0))
}
